### PR TITLE
Missing/incorrect exception handling and error message for validity check of email address format #71

### LINKED ISSUE #71: Missing/incorrect exception handling and error message for validity check of email address format
I noticed that in the current version of AB3, some invalid email addresses are not detected and can be added to a person without any error message. 
(Reference to valid email address format: https://help.xmatters.com/ondemand/trial/valid_email_format.htm)

The current constraint for email address format is as followed (it is only shown in error message, should be added into UG for clarification):
"Emails should be of the format local-part@domain and adhere to the following constraints:
1. The local-part should only contain alphanumeric characters and these special characters, excluding the parentheses, (!#$%&'*+/=?`{|}~^.-) .
2. This is followed by a '@' and then a domain name. The domain name must:
    - be at least 2 characters long
    - start and end with alphanumeric characters
    - consist of alphanumeric characters, a period or a hyphen for the characters in between, if any."

Examples of these undetected invalid email address formats include:
1. Starting or ending with a special character. E.g. `abc-@mail.com` and `.abc@mail.com`
2. Consecutive special letters. E.g. `abc..def@mail.com`
3. Missing top-level domain. E.g. `abc.def@mail`
4. Domain shorter than 2 characters. E.g. `abc.def@mail.c`

Example of wrongly rejected email address format:
1. Domain with more than 1 dot (every dot followed by at least two characters from a to z). E.g. `e123456@u.nus.edu`

I think that it will be better if the email address format accepted in AB3 can be modified accordingly to detect the better account for actual email addresses in reality.

### PR DESCRIPTION
**Pull request by siangernlow**
Fixes #71

### CODE REVIEW

**Comment by damithc** [comment]
Thanks for the PR @siangernlow 
Let's wait to see if some others can help with the reviews.
In the meantime, you might want to fix the unit tests in the same commit as the regex, as we expect each commit to be able to pass CI.

**Comment by damithc** [comment]
@se-edu/tech-team-level1 any reviews for this PR?

File: src/main/java/seedu/address/model/person/Email.java [hunk 1]
```diff
@@ -9,23 +9,27 @@
...
-    private static final String DOMAIN_FIRST_CHARACTER_REGEX = "[^\\W_]"; // alphanumeric characters except underscore
-    private static final String DOMAIN_MIDDLE_REGEX = "[a-zA-Z0-9.-]*"; // alphanumeric, period and hyphen
-    private static final String DOMAIN_LAST_CHARACTER_REGEX = "[^\\W_]$";
-    public static final String VALIDATION_REGEX = LOCAL_PART_REGEX + "@"
-            + DOMAIN_FIRST_CHARACTER_REGEX + DOMAIN_MIDDLE_REGEX + DOMAIN_LAST_CHARACTER_REGEX;
+    private static final String ALPHANUMERIC_NO_UNDERSCORE = "[^\\W_]"; // alphanumeric characters except underscore
+    private static final String LOCAL_PART_REGEX = "^" + ALPHANUMERIC_NO_UNDERSCORE + "+([" + SPECIAL_CHARACTERS + "]"
+            + ALPHANUMERIC_NO_UNDERSCORE + "+)*";
+    private static final String DOMAIN_PART_REGEX = ALPHANUMERIC_NO_UNDERSCORE
+            + "+(-" + ALPHANUMERIC_NO_UNDERSCORE + "+)*";
```
Comment (JinHao-L) [inline src/main/java/seedu/address/model/person/Email.java]: Should we put the (one or more character) `+` at the end of the `ALPHANUMERIC_NO_UNDERSCORE` instead? 

This way we can remove the '+' signs in `LOCAL_PART_REGEX` and `DOMAIN_PART_REGEX` which makes it easier to read and follow.

```suggestion
    private static final String ALPHANUMERIC_NO_UNDERSCORE = "[^\\W_]+"; // alphanumeric characters except underscore
    private static final String LOCAL_PART_REGEX = "^" + ALPHANUMERIC_NO_UNDERSCORE + "([" + SPECIAL_CHARACTERS + "]"
            + ALPHANUMERIC_NO_UNDERSCORE + ")*";
    private static final String DOMAIN_PART_REGEX = ALPHANUMERIC_NO_UNDERSCORE
            + "(-" + ALPHANUMERIC_NO_UNDERSCORE + ")*";
```

**Review by JinHao-L — COMMENTED** [review]
Other than the cosmetic change suggested. The rest looks good to me. 👍

**Comment by siangernlow** [comment]
Thanks for the review, @JinHao-L . The suggested changes have been implemented

**Comment by damithc** [comment]
Sorry @siangernlow overlooked this due to start-of-semester rush of work
@se-edu/tech-team-level1 any reviews for this?

**Review by JinHao-L — COMMENTED** [review]
The changes look good and the tests are pretty robust. Looks good to merge! 👍

**Comment by damithc** [comment]
Thanks for the fix @siangernlow 
I've combined the two commits (as changing the code without changing the message introduces an incorrect state into the code)

**Pull request merged**